presentation/src/main/java/com/exallium/mvvmcleanexample/presentation/users/UserEditViewModel.kt

```diff
@@ -1,61 +1,89 @@
+import com.exallium.mvvmcleanexample.presentation.utils.behaviourRelay
+import com.exallium.mvvmcleanexample.presentation.utils.bindable
+import com.exallium.mvvmcleanexample.presentation.utils.delegates.EcoDisposableProperty
+import com.exallium.mvvmcleanexample.presentation.utils.delegates.RelayProperty
+import com.exallium.mvvmcleanexample.presentation.utils.ecoDisposable
+import com.exallium.mvvmcleanexample.presentation.utils.notifiable.BaseNotifiableObservable
+import com.exallium.mvvmcleanexample.presentation.utils.notifiable.NotifiableObservable
+import com.exallium.mvvmcleanexample.presentation.utils.publishRelay
 
-class UserEditViewModel(userEditModel: UserEditModel,
-                        private val router: UiRouter) {
+class UserEditViewModel(private val userEditModel: UserEditModel,
+                        private val router: UiRouter,
+                        private val savedUserRelay: RelayProperty<User> = publishRelay<User>(),
+                        private val firstNameRelay: RelayProperty<String> = behaviourRelay<String>(),
+                        private val lastNameRelay: RelayProperty<String> = behaviourRelay<String>(),
+                        ecoDisposable: EcoDisposableProperty = ecoDisposable(null),
+                        notifiableObservable: NotifiableObservable = BaseNotifiableObservable())
+    : NotifiableObservable by notifiableObservable {
 
-    val firstNameError = ObservableField<String>("")
-    val lastNameError = ObservableField<String>("")
+    private var disposable by ecoDisposable
+    private var savedUser by savedUserRelay
+    private var firstName by firstNameRelay
+    private var lastName by lastNameRelay
 
-    private val compositeDisposable = CompositeDisposable()
+    init {
+        delegateTo(this)
+    }
 
-    private val clickRelay = PublishRelay.create<User>()
-    private val firstNameRelay = BehaviorRelay.create<String>()
-    private val lastNameRelay = BehaviorRelay.create<String>()
+    // region lifecycle
 
-    init {
-        compositeDisposable.add(userEditModel.results(clickRelay, firstNameRelay, lastNameRelay)
-                .subscribe(this::handleResult, this::handleError))
+    fun start() {
+        disposable = userEditModel.results(
+                saveClicks = savedUserRelay.relay,
+                firstNameChanges = firstNameRelay.relay,
+                lastNameChanges = lastNameRelay.relay)
+                .subscribe(this::handleResult, this::handleError)
     }
 
-    fun setSaveClicks(obs: Observable<Unit>) {
-        compositeDisposable.add(obs
-                .map { User(firstNameRelay.value, lastNameRelay.value) }
-                .subscribe(clickRelay))
+    fun stop() {
+        disposable = null
     }
 
-    fun setFirstNameChanged(obs: Observable<String>) {
-        compositeDisposable.add(obs.subscribe(firstNameRelay))
+    //endregion
+
+    //region bindable properties ( view << view model )
```

Comment:
agreed, it's hard to see where regular Kotlin properties fit into this model though.